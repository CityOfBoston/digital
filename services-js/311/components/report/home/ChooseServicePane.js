// @flow

import React from 'react';
import Head from 'next/head';
import Link from 'next/link';

import type { ServiceSummary } from '../../../data/types';

import SectionHeader from '../../common/SectionHeader';
import DescriptionBox from '../../common/DescriptionBox';

export type Props = {|
  description: string,
  handleDescriptionChanged: (ev: SyntheticInputEvent) => mixed,
  suggestedServiceSummaries: ?ServiceSummary[],
|};

function renderSummaryRow({ code, name, description }: { code: string, name: string, description: ?string }) {
  return (
    <div className="dr" key={code}>
      <Link href={`/report?code=${code}`} as={`/report/${code}`}><a className="dr-h">
        <div className="dr-ic" style={{ transform: 'translateY(-49%) rotateZ(-90deg)' }}><svg xmlns="http://www.w3.org/2000/svg" viewBox="-2 8.5 18 25"><path className="dr-i" d="M16 21L.5 33.2c-.6.5-1.5.4-2.2-.2-.5-.6-.4-1.6.2-2l12.6-10-12.6-10c-.6-.5-.7-1.5-.2-2s1.5-.7 2.2-.2L16 21z" /></svg></div>
        <div className="dr-t">{name}</div>
        <div className="dr-st">{description}</div>
      </a></Link>
    </div>
  );
}

function renderGeneralRequestRow() {
  return renderSummaryRow({
    code: 'BOS311GEN',
    name: 'General Request',
    // TODO(finh): Add something here
    description: '',
  });
}

function renderLoading() {
  return (
    <div className="t--info m-v300">
      Matching your problem to BOS:311 services…
    </div>
  );
}

function renderTypeDescription() {
  return (
    <div className="t--info m-v300">
      Type in your problem above to get matched with BOS:311 services.
    </div>
  );
}

function renderSuggestions(suggestedServiceSummaries: ServiceSummary[]) {
  return (
    <div>
      <div className="t--info m-v300">
        We’ve matched these services to your problem. Pick one to continue.
      </div>

      <div className="m-v500">
        { suggestedServiceSummaries.map(renderSummaryRow) }
      </div>

      <div className="t--info m-v300">
        If none of those seem like a good fit, you can submit a General Request.
      </div>

      <div className="m-v500">{renderGeneralRequestRow()}</div>
    </div>
  );
}

function renderNoSuggestions() {
  return (
    <div>
      <div className="t--info m-v300">
        We weren’t able to automatically match your problem with a service.
        File a General Request and someone will help you out.
      </div>

      <div className="m-v500">{renderGeneralRequestRow()}</div>
    </div>
  );
}

export default function ChooseServicePane({ description, handleDescriptionChanged, suggestedServiceSummaries }: Props) {
  return (
    <div>
      <Head>
        <title>BOS:311 — Choose a Service</title>
      </Head>

      <div className="p-a300 p-a800--xl" style={{ paddingBottom: '.75rem' }}>

        <SectionHeader>File a Report</SectionHeader>

        <div className="m-v500">
          <DescriptionBox
            minHeight={90}
            maxHeight={222}
            text={description}
            placeholder="Tell us your problem"
            onInput={handleDescriptionChanged}
          />
        </div>
      </div>

      <div className="b b--g m-v500 p-a300 p-a800--xl">
        <h3 className="stp">
          How can we help?
        </h3>

        { !suggestedServiceSummaries && renderLoading() }
        { suggestedServiceSummaries && suggestedServiceSummaries.length > 0 && renderSuggestions(suggestedServiceSummaries) }
        { suggestedServiceSummaries && suggestedServiceSummaries.length === 0 && description.length === 0 && renderTypeDescription() }
        { suggestedServiceSummaries && suggestedServiceSummaries.length === 0 && description.length > 0 && renderNoSuggestions() }
      </div>
    </div>
  );
}
