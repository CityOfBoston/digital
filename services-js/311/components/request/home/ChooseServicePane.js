// @flow

import * as React from 'react';
import Head from 'next/head';
import Link from 'next/link';
import { css } from 'glamor';

import type { ServiceSummary } from '../../../data/types';
import type Ui from '../../../data/store/Ui';

import SectionHeader from '../../common/SectionHeader';
import LoadingIcons from '../../common/LoadingIcons';

export type Props = {|
  description: string,
  suggestedServiceSummaries: ?(ServiceSummary[]),
  ui: Ui,
|};

const LOADING_INDICATORS_STYLE = css({
  overflow: 'hidden',
});

const LOADING_INDICATOR_WRAPPER_STYLE = css({
  height: 120,
  display: 'flex',
  flexDirection: 'column',
});

function renderSummaryRow(
  problemDescription: string,
  {
    code,
    name,
    description,
  }: { code: string, name: string, description: ?string }
) {
  return (
    <div className="dr" key={code}>
      <Link
        href={`/request?code=${code}&description=${encodeURIComponent(
          problemDescription
        )}`}
        as={`/request/${code}`}
      >
        <a className="dr-h">
          <div
            className="dr-ic"
            style={{ transform: 'translateY(-49%) rotateZ(-90deg)' }}
          >
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="-2 8.5 18 25">
              <path
                className="dr-i"
                d="M16 21L.5 33.2c-.6.5-1.5.4-2.2-.2-.5-.6-.4-1.6.2-2l12.6-10-12.6-10c-.6-.5-.7-1.5-.2-2s1.5-.7 2.2-.2L16 21z"
              />
            </svg>
          </div>
          <div className="dr-t">
            {name}
          </div>
          <div className="dr-st">
            <span
              className="t--info"
              style={{ textTransform: 'none', fontStyle: 'normal' }}
            >
              {description}
            </span>
          </div>
        </a>
      </Link>
    </div>
  );
}

function renderGeneralRequestRow(problemDescription: string) {
  return renderSummaryRow(problemDescription, {
    code: 'BOS311GEN',
    name: 'General Request',
    // TODO(finh): Add something here
    description: '',
  });
}

function renderLoading(ui: Ui) {
  return (
    <div>
      <div className="t--info">Matching your request to BOS:311 services…</div>

      <div className={`p-a300 g ${LOADING_INDICATORS_STYLE.toString()}`}>
        <div className={`g--4 ${LOADING_INDICATOR_WRAPPER_STYLE.toString()}`}>
          <LoadingIcons initialDelay={0} reduceMotion={ui.reduceMotion} />
        </div>
        <div className={`g--4 ${LOADING_INDICATOR_WRAPPER_STYLE.toString()}`}>
          <LoadingIcons initialDelay={100} reduceMotion={ui.reduceMotion} />
        </div>
        <div className={`g--4 ${LOADING_INDICATOR_WRAPPER_STYLE.toString()}`}>
          <LoadingIcons initialDelay={200} reduceMotion={ui.reduceMotion} />
        </div>
      </div>
    </div>
  );
}

function renderSuggestions(
  problemDescription: string,
  suggestedServiceSummaries: ServiceSummary[]
) {
  return (
    <div>
      <div className="t--info">
        We’ve matched these services to your request. Pick one to continue.
      </div>

      <div className="m-v500">
        {suggestedServiceSummaries.map(s =>
          renderSummaryRow(problemDescription, s)
        )}
      </div>

      <div className="t--info m-v300">
        If none of those seem like a good fit, you can submit a General Request.
      </div>

      <div className="m-v500">
        {renderGeneralRequestRow(problemDescription)}
      </div>
    </div>
  );
}

function renderNoSuggestions(problemDescription: string) {
  return (
    <div>
      <div className="t--info m-v300">
        We weren’t able to automatically match your request with a service. File
        a General Request and someone will help you out.
      </div>

      <div className="m-v500">
        {renderGeneralRequestRow(problemDescription)}
      </div>
    </div>
  );
}

export default function ChooseServicePane({
  description,
  suggestedServiceSummaries,
  ui,
}: Props) {
  return (
    <div>
      <Head>
        <title>BOS:311 — Choose a Service</title>
      </Head>

      <div className="p-a300 p-a800--xl" style={{ paddingBottom: '.75rem' }}>
        <SectionHeader>BOS:311 — Choose a Service</SectionHeader>

        {description &&
          <div className="m-v500 t--intro">
            “{description}”
          </div>}
      </div>

      <div className="b b--g p-a300 p-a800--xl">
        {!suggestedServiceSummaries && renderLoading(ui)}
        {suggestedServiceSummaries &&
          suggestedServiceSummaries.length > 0 &&
          renderSuggestions(description, suggestedServiceSummaries)}
        {suggestedServiceSummaries &&
          suggestedServiceSummaries.length === 0 &&
          renderNoSuggestions(description)}
      </div>
    </div>
  );
}
