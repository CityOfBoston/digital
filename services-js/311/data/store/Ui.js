// @flow

import { observable, action, computed, reaction } from 'mobx';
import throttle from 'lodash/throttle';

import { HEADER_HEIGHT } from '../../components/style-constants';

// Store to track window and scroll information.
export default class Ui {
  // This tracks the current window height, minus the fixed header height
  @observable visibleHeight: number = Number.MAX_SAFE_INTEGER;
  @observable visibleWidth: ?number;

  @observable scrollY: number = 0;
  @observable debouncedScrollY: number = 0;

  @observable visible: boolean = true;
  @observable reduceMotion: boolean = false;

  debouncedScrollDisposer: ?Function;

  @action
  attach() {
    this.debouncedScrollDisposer = reaction(
      () => this.scrollY,
      (scrollY) => { this.debouncedScrollY = scrollY; },
      {
        name: 'debounced scroll',
        delay: 250,
      },
    );

    if (typeof window === 'undefined') {
      return;
    }

    if (window.matchMedia) {
      this.reduceMotion = window.matchMedia('(prefers-reduced-motion)').matches;
    }

    window.addEventListener('scroll', this.handleScroll);
    window.addEventListener('resize', this.handleResize);

    this.visible = !document.hidden;
    document.addEventListener('visibilitychange', this.handleVisibility);

    this.handleScroll();
    this.handleResize();
  }

  detach() {
    if (this.debouncedScrollDisposer) {
      this.debouncedScrollDisposer();
      this.debouncedScrollDisposer = null;
    }

    if (typeof window === 'undefined') {
      return;
    }

    window.removeEventListener('scroll', this.handleScroll);
    window.removeEventListener('resize', this.handleResize);

    document.removeEventListener('visibilitychange', this.handleVisibility);
  }

  handleScroll = throttle(action('scroll handler', () => {
    this.scrollY = window.scrollY || (document.documentElement && document.documentElement.scrollTop) || 0;
  }), 100);

  handleResize = throttle(action('resize handler', () => {
    this.visibleHeight = window.innerHeight - HEADER_HEIGHT;
    this.visibleWidth = window.innerWidth;
  }), 100);

  @action.bound
  handleVisibility() {
    this.visible = !document.hidden;
  }

  @computed get mediaSmall(): boolean { return !!this.visibleWidth && this.visibleWidth > 480; }
  @computed get mediaMedium(): boolean { return !!this.visibleWidth && this.visibleWidth > 768; }
  @computed get mediaLarge(): boolean { return !!this.visibleWidth && this.visibleWidth > 840; }
  @computed get mediaXLarge(): boolean { return !!this.visibleWidth && this.visibleWidth > 980; }
  @computed get mediaXXLarge(): boolean { return !!this.visibleWidth && this.visibleWidth > 1280; }

  // true if less that mediaLarge, but false if we haven't determined the screen size yet
  @computed get belowMediaLarge(): boolean { return !!this.visibleWidth && !this.mediaLarge; }
}
