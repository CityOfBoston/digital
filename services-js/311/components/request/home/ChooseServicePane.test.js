// @flow

import React from 'react';
import renderer from 'react-test-renderer';
import Ui from '../../../data/store/Ui';
import ChooseServicePane from './ChooseServicePane';

const SERVICE_SUMMARIES = [{
  name: 'Cosmic Incursion',
  code: 'CSMCINC',
  description: 'Something is threatening the fabric of the universe',
  group: 'ultimate',
}];

describe('rendering', () => {
  let ui;

  beforeEach(() => {
    ui = new Ui();
  });

  test('some suggested summaries', () => {
    const component = renderer.create(
      <ChooseServicePane description="Thanos is attacking" suggestedServiceSummaries={SERVICE_SUMMARIES} ui={ui} />,
    );

    const json = component.toJSON();
    expect(json).toMatchSnapshot();
  });

  test('no suggested summaries', () => {
    const component = renderer.create(
      <ChooseServicePane description="Thanos is attacking" suggestedServiceSummaries={[]} ui={ui} />,
    );

    const json = component.toJSON();
    expect(json).toMatchSnapshot();
  });

  test('loading', () => {
    const component = renderer.create(
      <ChooseServicePane description="Thanos is attacking" suggestedServiceSummaries={null} ui={ui} />,
      { createNodeMock: () => document.createElement('div') },
    );

    const json = component.toJSON();
    expect(json).toMatchSnapshot();
  });

  test('no description', () => {
    const component = renderer.create(
      <ChooseServicePane description="" suggestedServiceSummaries={[]} ui={ui} />,
    );

    const json = component.toJSON();
    expect(json).toMatchSnapshot();
  });
});
