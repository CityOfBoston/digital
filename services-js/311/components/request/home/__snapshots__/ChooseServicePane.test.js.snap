// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering loading 1`] = `
<div>
  <div
    className="p-a300 p-a800--xl"
    style={
      Object {
        "paddingBottom": ".75rem",
      }
    }
  >
    <div
      className="sh sh--y"
    >
      <h1
        className="sh-title"
      >
        BOS:311 — Choose a Service
      </h1>
    </div>
    <div
      className="m-v500 t--intro"
    >
      “
      Thanos is attacking
      ”
    </div>
  </div>
  <div
    className="b b--g p-a300 p-a800--xl"
  >
    <div>
      <div
        className="t--info"
      >
        Matching your request to BOS:311 services…
      </div>
      <div
        className="p-a300 g css-rdsogp"
      >
        <div
          className="g--4 css-19eoixt"
        >
          <span
            className="css-17mox4x"
          >
            <svg
              className={
                Object {
                  "data-css-1acy7qc": "",
                }
              }
              dangerouslySetInnerHTML={
                Object {
                  "__html": "<use xlink:href=\\"/static/img/svg/loading-icons.svg#311_icons_general_lighting_request\\" height=\\"100%\\"></use>",
                }
              }
              role="img"
            />
          </span>
        </div>
        <div
          className="g--4 css-19eoixt"
        >
          <span
            className="css-17mox4x"
          >
            <svg
              className={
                Object {
                  "data-css-1acy7qc": "",
                }
              }
              dangerouslySetInnerHTML={
                Object {
                  "__html": "<use xlink:href=\\"/static/img/svg/loading-icons.svg#311_icons_general_lighting_request\\" height=\\"100%\\"></use>",
                }
              }
              role="img"
            />
          </span>
        </div>
        <div
          className="g--4 css-19eoixt"
        >
          <span
            className="css-17mox4x"
          >
            <svg
              className={
                Object {
                  "data-css-1acy7qc": "",
                }
              }
              dangerouslySetInnerHTML={
                Object {
                  "__html": "<use xlink:href=\\"/static/img/svg/loading-icons.svg#311_icons_general_lighting_request\\" height=\\"100%\\"></use>",
                }
              }
              role="img"
            />
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`rendering no description 1`] = `
<div>
  <div
    className="p-a300 p-a800--xl"
    style={
      Object {
        "paddingBottom": ".75rem",
      }
    }
  >
    <div
      className="sh sh--y"
    >
      <h1
        className="sh-title"
      >
        BOS:311 — Choose a Service
      </h1>
    </div>
    
  </div>
  <div
    className="b b--g p-a300 p-a800--xl"
  >
    <div>
      <div
        className="t--info m-v300"
      >
        We weren’t able to automatically match your request with a service. File a General Request and someone will help you out.
      </div>
      <div
        className="m-v500"
      >
        <div
          className="dr"
        >
          <a
            className="dr-h"
            href="/request/BOS311GEN"
            onClick={[Function]}
          >
            <div
              className="dr-ic"
              style={
                Object {
                  "transform": "translateY(-49%) rotateZ(-90deg)",
                }
              }
            >
              <svg
                viewBox="-2 8.5 18 25"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  className="dr-i"
                  d="M16 21L.5 33.2c-.6.5-1.5.4-2.2-.2-.5-.6-.4-1.6.2-2l12.6-10-12.6-10c-.6-.5-.7-1.5-.2-2s1.5-.7 2.2-.2L16 21z"
                />
              </svg>
            </div>
            <div
              className="dr-t"
            >
              General Request
            </div>
            <div
              className="dr-st"
            >
              <span
                className="t--info"
                style={
                  Object {
                    "fontStyle": "normal",
                    "textTransform": "none",
                  }
                }
              >
                
              </span>
            </div>
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`rendering no suggested summaries 1`] = `
<div>
  <div
    className="p-a300 p-a800--xl"
    style={
      Object {
        "paddingBottom": ".75rem",
      }
    }
  >
    <div
      className="sh sh--y"
    >
      <h1
        className="sh-title"
      >
        BOS:311 — Choose a Service
      </h1>
    </div>
    <div
      className="m-v500 t--intro"
    >
      “
      Thanos is attacking
      ”
    </div>
  </div>
  <div
    className="b b--g p-a300 p-a800--xl"
  >
    <div>
      <div
        className="t--info m-v300"
      >
        We weren’t able to automatically match your request with a service. File a General Request and someone will help you out.
      </div>
      <div
        className="m-v500"
      >
        <div
          className="dr"
        >
          <a
            className="dr-h"
            href="/request/BOS311GEN"
            onClick={[Function]}
          >
            <div
              className="dr-ic"
              style={
                Object {
                  "transform": "translateY(-49%) rotateZ(-90deg)",
                }
              }
            >
              <svg
                viewBox="-2 8.5 18 25"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  className="dr-i"
                  d="M16 21L.5 33.2c-.6.5-1.5.4-2.2-.2-.5-.6-.4-1.6.2-2l12.6-10-12.6-10c-.6-.5-.7-1.5-.2-2s1.5-.7 2.2-.2L16 21z"
                />
              </svg>
            </div>
            <div
              className="dr-t"
            >
              General Request
            </div>
            <div
              className="dr-st"
            >
              <span
                className="t--info"
                style={
                  Object {
                    "fontStyle": "normal",
                    "textTransform": "none",
                  }
                }
              >
                
              </span>
            </div>
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`rendering some suggested summaries 1`] = `
<div>
  <div
    className="p-a300 p-a800--xl"
    style={
      Object {
        "paddingBottom": ".75rem",
      }
    }
  >
    <div
      className="sh sh--y"
    >
      <h1
        className="sh-title"
      >
        BOS:311 — Choose a Service
      </h1>
    </div>
    <div
      className="m-v500 t--intro"
    >
      “
      Thanos is attacking
      ”
    </div>
  </div>
  <div
    className="b b--g p-a300 p-a800--xl"
  >
    <div>
      <div
        className="t--info"
      >
        We’ve matched these services to your request. Pick one to continue.
      </div>
      <div
        className="m-v500"
      >
        <div
          className="dr"
        >
          <a
            className="dr-h"
            href="/request/CSMCINC"
            onClick={[Function]}
          >
            <div
              className="dr-ic"
              style={
                Object {
                  "transform": "translateY(-49%) rotateZ(-90deg)",
                }
              }
            >
              <svg
                viewBox="-2 8.5 18 25"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  className="dr-i"
                  d="M16 21L.5 33.2c-.6.5-1.5.4-2.2-.2-.5-.6-.4-1.6.2-2l12.6-10-12.6-10c-.6-.5-.7-1.5-.2-2s1.5-.7 2.2-.2L16 21z"
                />
              </svg>
            </div>
            <div
              className="dr-t"
            >
              Cosmic Incursion
            </div>
            <div
              className="dr-st"
            >
              <span
                className="t--info"
                style={
                  Object {
                    "fontStyle": "normal",
                    "textTransform": "none",
                  }
                }
              >
                Something is threatening the fabric of the universe
              </span>
            </div>
          </a>
        </div>
      </div>
      <div
        className="t--info m-v300"
      >
        If none of those seem like a good fit, you can submit a General Request.
      </div>
      <div
        className="m-v500"
      >
        <div
          className="dr"
        >
          <a
            className="dr-h"
            href="/request/BOS311GEN"
            onClick={[Function]}
          >
            <div
              className="dr-ic"
              style={
                Object {
                  "transform": "translateY(-49%) rotateZ(-90deg)",
                }
              }
            >
              <svg
                viewBox="-2 8.5 18 25"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  className="dr-i"
                  d="M16 21L.5 33.2c-.6.5-1.5.4-2.2-.2-.5-.6-.4-1.6.2-2l12.6-10-12.6-10c-.6-.5-.7-1.5-.2-2s1.5-.7 2.2-.2L16 21z"
                />
              </svg>
            </div>
            <div
              className="dr-t"
            >
              General Request
            </div>
            <div
              className="dr-st"
            >
              <span
                className="t--info"
                style={
                  Object {
                    "fontStyle": "normal",
                    "textTransform": "none",
                  }
                }
              >
                
              </span>
            </div>
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;
